README.md
# UpMarkdown

<!-- This is the README for your extension "upmarkdown". After writing up a brief description, we recommend including the following sections.

## Features

Describe specific features of your extension including screenshots of your extension in action. Image paths are relative to this README file.

For example if there is an image subfolder under your extension project workspace:

\!\[feature X\]\(images/feature-x.png\)

> Tip: Many popular extensions utilize animations. This is an excellent way to show off your extension! We recommend short, focused animations that are easy to follow.

## Requirements

If you have any requirements or dependencies, add a section describing those and how to install and configure them.

## Extension Settings

Include if your extension adds any VS Code settings through the `contributes.configuration` extension point.

For example:

This extension contributes the following settings:

* `myExtension.enable`: enable/disable this extension
* `myExtension.thing`: set to `blah` to do something

## Known Issues

Calling out known issues can help limit users opening duplicate issues against your extension.

## Release Notes

Users appreciate release notes as you update your extension.

### 1.0.0

Initial release of ...

### 1.0.1

Fixed issue #.

### 1.1.0

Added features X, Y, and Z. -->


CHANGELOG.md
# Change Log

All notable changes to the "upmarkdown" extension will be documented in this file.

Check [Keep a Changelog](http://keepachangelog.com/) for recommendations on how to structure this file.

## [Unreleased]

- Initial release

src/extension.ts
import * as vscode from 'vscode';
// import * as path from 'path';

// this method is called when your extension is activated
// your extension is activated the very first time the command is executed
export function activate(context: vscode.ExtensionContext) {

	// Use the console to output diagnostic information (console.log) and errors (console.error)
	// This line of code will only be executed once when your extension is activated
	// console.log('Congratulations, your extension "upmarkdown" is now active!');
	context.globalState.update("test", undefined);

	// The command has been defined in the package.json file
	// Now provide the implementation of the command with registerCommand
	// The commandId parameter must match the command field in package.json
	let disposable = vscode.commands.registerCommand('extension.helloWorld', () => {
		const val: string = context.globalState.get("test", "defaultValue");

		// const folderName = path.dirname(e.path);
		// const folderUrl = vscode.Uri.file(folderName);

		// Display a message box to the user
		vscode.window.showInformationMessage(val);
	});

	context.subscriptions.push(disposable);
}

// this method is called when your extension is deactivated
export function deactivate() { }


src/_testFileStructureFunctionality/testFiles.js
"use strict";
// 1. select mother directory and assign Path var to the name of the mother directory
exports.__esModule = true;
// 2. recursively look for other directories, if exists go in and add name of directory to Path var
// 3. if not directory, assign name of file as key with the value as an object with Path var and location in file of all links to update
// (what if two files in different locations have the same name?)
/* example fs:
dir1/file1
dir1/file2
dir1/dir2
dir1/dir2/file3
dir1/dir2/dir3
dir1/dir2/dir3/file4

- dir1 is mother directory, Path = "dir1"
- file1 is saved as file1 = {path: path, links: {fileName: (position, length)}}
file1.path -> "dir1"
file1.links.file2 -> (600, 32)
*/
// 4. when done saving fs into an object/database, go into each file and update the links in the text. 
// *if using relative paths instead of absolute, another step after this one will require to update the length value of each link reference*
/* example fs if file1 has references to both file2 and file3:
dir1/file1
dir1/file2
dir1/dir2/file3

file2 moved to dir2:
dir1/file1
dir1/dir2/file2
dir1/dir2/file3

first, scan and update the fs object:
file2.path updates from "dir1" to "dir1/dir2", based on Path var
file2.links stays the same
file1.path stays the same
file1.links.file2 stays the same

then, update links:
file2.path & file2.links stays the same
file1.path stays the same
file1 link to file2 in the text is updated based on new path of file2 minus current path
*file1.links.file2 is updated with the new length

*length doesn't need to be updated if only using absolute paths*
*/
var fs = require("fs");
var path = require("path");
fs.readdir('.', function (err, files) {
    var rootPath = path.resolve(__dirname).replace(/.*\//, '');
    console.log(rootPath);
    if (err) {
        throw err;
    }
    for (var i in files) {
        if (fs.existsSync(files[i]) && fs.lstatSync(files[i]).isDirectory()) {
            console.log(files[i]);
        }
    }
});


src/_testFileStructureFunctionality/testFiles.ts
// 1. select mother directory and assign Path var to the name of the mother directory

// 2. recursively look for other directories, if exists go in and add name of directory to Path var

// 3. if not directory, assign name of file as key with the value as an object with Path var and location in file of all links to update
// (what if two files in different locations have the same name?)

/* example fs:
dir1/file1
dir1/file2
dir1/dir2
dir1/dir2/file3
dir1/dir2/dir3
dir1/dir2/dir3/file4

- dir1 is mother directory, Path = "dir1"
- file1 is saved as file1 = {path: path, links: {fileName: (position, length)}}
file1.path -> "dir1"
file1.links.file2 -> (600, 32)
*/

// 4. when done saving fs into an object/database, go into each file and update the links in the text. 
// *if using relative paths instead of absolute, another step after this one will require to update the length value of each link reference*

/* example fs if file1 has references to both file2 and file3:
dir1/file1
dir1/file2
dir1/dir2/file3

file2 moved to dir2:
dir1/file1
dir1/dir2/file2
dir1/dir2/file3

first, scan and update the fs object:
file2.path updates from "dir1" to "dir1/dir2", based on Path var
file2.links stays the same
file1.path stays the same
file1.links.file2 stays the same

then, update links:
file2.path & file2.links stays the same
file1.path stays the same
file1 link to file2 in the text is updated based on new path of file2 minus current path
*file1.links.file2 is updated with the new length

*length doesn't need to be updated if only using absolute paths*
*/

import * as fs from 'fs';
import * as path from 'path';

interface IPaths {
  [prop: string]: IPath; //fileName: file props
}

interface IPath {
  path: string;
  links: {
  } & {
    [prop: string]: [number, number], //fileName: [location, length]
  };
}

const getFS = (): IPaths => {

  return {
    'file1': { path: '/path', links: { 'name-of-file': [50, 12] } },
    'file2': { path: '/path2', links: { 'name-of-file': [100, 12] } }
  };
};

fs.readdir('.', function (err, files) {
  const currentPath = path.resolve(__dirname).replace(/.*\//, '');
  if (err) { throw err; }
  for (let i in files) {
    if (fs.existsSync(files[i]) && fs.lstatSync(files[i]).isDirectory()) { // if current file is a directory
      console.log(files[i]);
    }
  }
});